File: angry-children-2/code1.pl
#!/usr/bin/perl

use strict;
use warnings;

# Complete the angryChildren function below.
sub angryChildren {
    my ($k, $packets) = @_;

    my @ordered_packets = sort {$a <=> $b} @$packets;

    my $cdiff = 0;

    for (my $i=0; $i<$k; ++$i) {
        my $y = $ordered_packets[$i];
        foreach my $item (@ordered_packets[0 .. $i-1]) {
            if ($item < $y) {
                $cdiff += $y - $item;
            } else {
                $cdiff += $item - $y;
            }
        }
    }

    my $min_cdiff = $cdiff;

    for (my $i=$k; $i<@ordered_packets; ++$i) {
        my $x = $ordered_packets[$i-$k];
        my $y = $ordered_packets[$i];
        my ($min, $max) = $x<$y ? ($x, $y) : ($y, $x);
        my $y_minus_x = $y - $x;
        my $x_minus_y = $x - $y;
        my $x_plus_y  = $x + $y;
        foreach my $item (@ordered_packets[$i-$k+1 .. $i-1]) {
            if ($item < $min) {
                $cdiff += $y_minus_x;
            } elsif ($item > $max) {
                $cdiff += $x_minus_y;
            } elsif ($x < $y) {
                $cdiff += $x_plus_y - 2*$item;
            } else {
                $cdiff += 2*$item - $x_plus_y;
            }
        }
        $min_cdiff = $cdiff if $cdiff < $min_cdiff;
    }

    return $min_cdiff;
}

open(my $fptr, '>', $ENV{'OUTPUT_PATH'});
$fptr = *STDOUT;

my $n = <>;
$n =~ s/\s+$//;

my $k = <>;
$k =~ s/\s+$//;

my @packets = ();

for (1..$n) {
    my $packets_item = <>;
    $packets_item =~ s/\s+$//;
    push @packets, $packets_item;
}

my $result = angryChildren $k, \@packets;

print $fptr "$result\n";

close $fptr;

File: angry-children-2/code1.cc
#include <bits/stdc++.h>

using namespace std;

// Complete the angryChildren function below.
long angryChildren(int k, vector<int> packets) {
    sort(packets.begin(), packets.end());

    long cdiff = 0;
    
    for (int i=0; i<k; ++i) {
        int y = packets[i];
        for (vector<int>::iterator it=packets.begin(); it != packets.begin()+i; ++it) {
            if (*it < y)
                cdiff += y - *it;
            else
                cdiff += *it - y;
        }
    }

    long min_cdiff = cdiff;

    for (int i=k; i<packets.size(); ++i) {
        int x = packets[i-k];
        int y = packets[i];
        int min = x<y ? x : y;
        int max = x>y ? x : y;
        for (vector<int>::iterator it=packets.begin()+i-k+1; it != packets.begin()+i; ++it) {
            if (*it < min)
                cdiff += y - x;
            else if (*it > max)
                cdiff += x - y;
            else if (x < y)
                cdiff += x + y - 2*(*it);
            else
                cdiff += 2*(*it) - x - y;
        }
        if (cdiff < min_cdiff)
            min_cdiff = cdiff;
    }

    return min_cdiff;
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    int n;
    cin >> n;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    int k;
    cin >> k;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    vector<int> packets(n);

    for (int i = 0; i < n; i++) {
        int packets_item;
        cin >> packets_item;
        cin.ignore(numeric_limits<streamsize>::max(), '\n');

        packets[i] = packets_item;
    }

    long result = angryChildren(k, packets);

    fout << result << "\n";

    fout.close();

    return 0;
}
